### PR TITLE
added bash autocompletion for ldap sync config

### PR DESCRIPTION
**Pull request by stevekuznetsov**
fixes https://github.com/openshift/origin/issues/5019 and adds autocompletion for {white,black}list files as well

cc/ @deads2k

### CODE REVIEW

**Comment by stevekuznetsov** [comment]
Using the same approach as upstream on `oc create` for `-f` ... even the bash autocompletion output is generated to be the same, but this doesn't work for me. Unclear why.

**Comment by liggitt** [comment]
There's a helper function (see https://github.com/openshift/origin/blob/b7de777c90fd892cd746c905c17d92710eae3de8/pkg/cmd/server/start/start_controllers.go#L89)

Like you, the generated file name completions stopped working for me recently. On my list to investigate why

**Comment by stevekuznetsov** [comment]
I'll update to use that helper. Currently `oc create -f` (and variants like `oc process`) work fine for me. It was just this specific flag that didn't seem to ...

**Comment by stevekuznetsov** [comment]
With https://github.com/openshift/origin/pull/5714, hopefully this will make these completions work. The two bugs are unrelated, however, so I think this PR is ready.

File: pkg/cmd/experimental/syncgroups/cli/syncgroups.go [hunk 1]
```diff
@@ -143,10 +143,16 @@ func NewCmdSyncGroups(name, fullName string, f *clientcmd.Factory, out io.Writer
...
 
 	cmd.Flags().StringVar(&whitelistFile, "whitelist", whitelistFile, "path to the group whitelist file")
+	cmd.MarkFlagFilename("whitelist", "txt")
 	cmd.Flags().StringVar(&blacklistFile, "blacklist", whitelistFile, "path to the group blacklist file")
+	cmd.MarkFlagFilename("blacklist", "txt")
 	// TODO enable this we're able to support string slice elements that have commas
 	// cmd.Flags().StringSliceVar(&options.Blacklist, "blacklist-group", options.Blacklist, "group to blacklist")
+
 	cmd.Flags().StringVar(&configFile, "sync-config", configFile, "path to the sync config")
+	cmd.MarkFlagFilename("sync-config", "json", "yaml", "yml")
```
Comment (liggitt) [inline pkg/cmd/experimental/syncgroups/cli/syncgroups.go]: I didn't realize we supported json sync configs today

Comment (stevekuznetsov) [inline pkg/cmd/experimental/syncgroups/cli/syncgroups.go]: That was scrapped. I'll amend.

**Comment by stevekuznetsov** [comment]
@liggitt I'm still not getting correct behavior here - `--whitelist` and `--blacklist` flags aren't auto-completing the names, whereas `--sync-config` is. Furthermore, only `--sync-config` auto-completes filenames, but does so correctly.

**Comment by deads2k** [comment]
> @liggitt I'm still not getting correct behavior here - --whitelist and --blacklist flags aren't auto-completing the names, whereas --sync-config is. Furthermore, only --sync-config auto-completes filenames, but does so correctly.

@fabianofranz any advice?

**Comment by liggitt** [comment]
do you have txt files it should be autocompleting?

**Comment by stevekuznetsov** [comment]
Yes. However, even the names of the flags aren't getting completed, *e.g.* `--white<tab><tab>`.

**Comment by stevekuznetsov** [comment]
`--sync-config` flag name auto completes correctly.

**Comment by fabianofranz** [comment]
You have a required flag, `--sync-config`, in which case our completions will always force you to provide it first. After you provide it you will get completions for everything else. Try `oadm groups prune --sync-config=/path/to/something --white<tab><tab>`. Adding @eparis who implemented this behaviour originally.

**Comment by fabianofranz** [comment]
Note that if you don't want your completions to behave like that, just remove marking `--sync-config` as required and rely on a validation.

**Comment by stevekuznetsov** [comment]
I see. I don't think anyone's going to stumble on this command and attempt to figure it out from CLI auto-completion help, so not marking it as required is probably fine. Thanks @fabianofranz

**Comment by stevekuznetsov** [comment]
@deads2k @liggitt if not having the flag marked as required is OK, this is ready.

**Comment by liggitt** [comment]
unmark wip, LGTM [merge]

**Comment by liggitt** [comment]
@fabianofranz that's a slightly bizarre behavior for required flag autocompletion

**Comment by eparis** [comment]
I could take another look, but if `--sync-config` is required what's wrong/bizarre with it being the only thing suggested until you set it...

**Comment by eparis** [comment]
I guess 'good' behavior would be that `--[tab][tab]` would only find `sync-config` but `--w[tab][tab]` would still find `whitelist`.  Not sure how to do it. Maybe if I get a minute I'll try to figure out how to implement that...

**Comment by deads2k** [comment]
> I guess 'good' behavior would be that --[tab][tab] would only find sync-config but --w[tab][tab] would still find whitelist. Not sure how to do it. Maybe if I get a minute I'll try to figure out how to implement that...

Yes, I would like that better.

**Comment by liggitt** [comment]
> I guess 'good' behavior would be that --[tab][tab] would only find sync-config but --w[tab][tab] would still find whitelist. Not sure how to do it. Maybe if I get a minute I'll try to figure out how to implement that...

+1. Letting validation catch the required flag isn't terrible until we get that, though, so I wouldn't consider it high priority.

**Comment by fabianofranz** [comment]
+1

**Comment by stevekuznetsov** [comment]
Flake on `test-end-to-end` [here](https://ci.openshift.redhat.com/jenkins/job/test_pull_requests_origin/8144/console):
```
!!! Error in hack/../test/end-to-end/core.sh:163
	'oc run cli-with-token-2 --attach --env=POD_NAMESPACE=test --image=openshift/origin:${TAG} --restart=Never -- cli whoami --loglevel=4 > ${LOG_DIR}/cli-with-token2.log 2>&1' exited with status 1
```

Apparently that pod [already existed](https://ci.openshift.redhat.com/jenkins/job/test_pull_requests_origin/8144/artifact/origin/artifacts/e2e/logs/cli-with-token2.log) when running the command. @liggitt I think this is a new flake.

**Comment by liggitt** [comment]
Same flake we've seen where an etcd server error response causes the client to retry a request which had actually been persisted server side, and therefore fails with an AlreadyExists or Conflict error

**Comment by stevekuznetsov** [comment]
@liggitt this needs another merge tag

**Pull request merged**